File: apps/ios/Packages/TrackEvolutionKit/Sources/TrackEvolutionKit/Telemetry/ParsedTelemetry.swift
import Foundation

/// One parsed telemetry source, whatever produced it.
///
/// The shape every parser in `public/js/import/parse.js` resolves to, and the
/// reason the review flow doesn't care where laps came from: a PDR file with
/// beacons arrives with exact laps and skips the line picker, a GoPro clip
/// arrives with a trace and `needsLine`, and a phone recording
/// (`ParsedRecording`) converts into the same value. NS-30 generalised
/// `ReviewModel`'s input to this so all three share one screen.
public struct ParsedTelemetry: Sendable {
    /// Which parser produced this. `KIND_LABELS` in the JS.
    public enum Kind: String, Sendable {
        case pdr
        case gopro
        /// Not a file parser: the in-app lap recorder.
        case live

        /// The label the default session name is built from.
        public var label: String {
            switch self {
            case .pdr: "PDR"
            case .gopro: "GoPro"
            case .live: "Recorded"
            }
        }
    }

    public var kind: Kind
    /// Local `yyyy-MM-dd` from the file, when it carries one.
    public var date: String?
    /// Local `HH:mm:ss`.
    public var time: String?
    public var durationS: Double
    public var laps: [ParsedLap]
    /// Decoded GPS trace in degrees, or nil when the source has none the app can
    /// use — normal for a PDR recording whose coordinates don't decode.
    public var gps: [Geo.Point]?
    /// The user has to pick a start/finish line before this has laps.
    public var needsLine: Bool
    /// Best-lap polyline, drawn as the racing line on the event page.
    public var bestLapTrace: [TracePoint]?
    /// Per-lap channel arrays on a driven-distance grid, stored with the session.
    public var lapChannels: SessionChannels?
    /// Session maxima from a PDR file's car channels.
    public var metrics: Metrics?

    // MARK: - PDR-only

    /// How many beacon crossings the recorder actually logged. Zero means every
    /// lap here was derived rather than timed.
    public var beaconCount: Int
    /// Raw latitude and odometer series, kept for lap recovery (`PDRLaps`).
    public var channels: RawChannels?
    /// Scaled car channels for the per-lap graphs.
    public var carChannels: CarChannels
    /// The slow channels, keyed by `SCALAR_NAMES`, before they are reduced to
    /// one value per lap. A dictionary rather than a struct because nothing
    /// reads an individual one — `buildLapChannels` walks the whole list.
    public var lapScalarChannels: [String: [ChannelPoint]]
    /// Session-level numbers, stored as the channel blob's `meta`.
    public var sessionMeta: SessionMeta?
    /// Set when laps came from lat-vs-distance periodicity rather than beacons.
    public var lapRecovery: LapRecovery?

    public init(
        kind: Kind,
        date: String? = nil,
        time: String? = nil,
        durationS: Double = 0,
        laps: [ParsedLap] = [],
        gps: [Geo.Point]? = nil,
        needsLine: Bool = false,
        bestLapTrace: [TracePoint]? = nil,
        lapChannels: SessionChannels? = nil,
        metrics: Metrics? = nil,
        beaconCount: Int = 0,
        channels: RawChannels? = nil,
        carChannels: CarChannels = CarChannels(),
        lapScalarChannels: [String: [ChannelPoint]] = [:],
        sessionMeta: SessionMeta? = nil,
        lapRecovery: LapRecovery? = nil
    ) {
        self.kind = kind
        self.date = date
        self.time = time
        self.durationS = durationS
        self.laps = laps
        self.gps = gps
        self.needsLine = needsLine
        self.bestLapTrace = bestLapTrace
        self.lapChannels = lapChannels
        self.metrics = metrics
        self.beaconCount = beaconCount
        self.channels = channels
        self.carChannels = carChannels
        self.lapScalarChannels = lapScalarChannels
        self.sessionMeta = sessionMeta
        self.lapRecovery = lapRecovery
    }

    /// Session maxima — each nil when the file doesn't carry the channel, or
    /// carries one whose peak isn't plausible.
    public struct Metrics: Hashable, Sendable {
        public var topSpeedKph: Double?
        public var maxRpm: Double?
        public var maxLatG: Double?
        /// Peak braking, reported positive the way a driver talks about it —
        /// it is the negative half of longitudinal G.
        public var maxBrakeG: Double?
        public var maxBoostKpa: Double?
        public var maxOilC: Double?

        public init(
            topSpeedKph: Double? = nil, maxRpm: Double? = nil, maxLatG: Double? = nil,
            maxBrakeG: Double? = nil, maxBoostKpa: Double? = nil, maxOilC: Double? = nil
        ) {
            self.topSpeedKph = topSpeedKph
            self.maxRpm = maxRpm
            self.maxLatG = maxLatG
            self.maxBrakeG = maxBrakeG
            self.maxBoostKpa = maxBoostKpa
            self.maxOilC = maxOilC
        }
    }

    /// The two raw PDR series lap recovery works from.
    public struct RawChannels: Hashable, Sendable {
        public var latPts: [ChannelPoint]
        public var odoPts: [ChannelPoint]

        public init(latPts: [ChannelPoint], odoPts: [ChannelPoint]) {
            self.latPts = latPts
            self.odoPts = odoPts
        }
    }

    /// Car channels in display units: speed km/h, latG in G, throttle/brake in %,
    /// steering in signed steering-wheel degrees. nil where the file lacks the
    /// channel.
    public struct CarChannels: Hashable, Sendable {
        public var speed: [ChannelPoint]?
        public var rpm: [ChannelPoint]?
        public var latG: [ChannelPoint]?
        public var throttle: [ChannelPoint]?
        public var brake: [ChannelPoint]?
        public var steering: [ChannelPoint]?
        /// Signed: negative under braking.
        public var longG: [ChannelPoint]?
        /// Degrees per second, signed.
        public var yaw: [ChannelPoint]?
        /// 1–8, or 0 for the clutch-in / no-gear state.
        public var gear: [ChannelPoint]?
        /// (driven − non-driven) wheelspeed, %.
        public var wheelSlip: [ChannelPoint]?
        /// Manifold gauge pressure, kPa.
        public var boost: [ChannelPoint]?
        /// ABS | traction control << 1 | stability control << 2.
        public var flags: [ChannelPoint]?

        public init(
            speed: [ChannelPoint]? = nil, rpm: [ChannelPoint]? = nil, latG: [ChannelPoint]? = nil,
            throttle: [ChannelPoint]? = nil, brake: [ChannelPoint]? = nil,
            steering: [ChannelPoint]? = nil, longG: [ChannelPoint]? = nil,
            yaw: [ChannelPoint]? = nil, gear: [ChannelPoint]? = nil,
            wheelSlip: [ChannelPoint]? = nil, boost: [ChannelPoint]? = nil,
            flags: [ChannelPoint]? = nil
        ) {
            self.speed = speed
            self.rpm = rpm
            self.latG = latG
            self.throttle = throttle
            self.brake = brake
            self.steering = steering
            self.longG = longG
            self.yaw = yaw
            self.gear = gear
            self.wheelSlip = wheelSlip
            self.boost = boost
            self.flags = flags
        }

        /// `chans[name]` in the JS, which indexes a plain object. Every caller
        /// walks `TelemetryChannels.CHANNEL_NAMES`, so keeping one lookup here
        /// is what stops a channel from being added to the list and silently
        /// never read.
        public subscript(name: String) -> [ChannelPoint]? {
            get {
                switch name {
                case "speed": speed
                case "rpm": rpm
                case "latG": latG
                case "throttle": throttle
                case "brake": brake
                case "steering": steering
                case "longG": longG
                case "yaw": yaw
                case "gear": gear
                case "wheelSlip": wheelSlip
                case "boost": boost
                case "flags": flags
                default: nil
                }
            }
            set {
                switch name {
                case "speed": speed = newValue
                case "rpm": rpm = newValue
                case "latG": latG = newValue
                case "throttle": throttle = newValue
                case "brake": brake = newValue
                case "steering": steering = newValue
                case "longG": longG = newValue
                case "yaw": yaw = newValue
                case "gear": gear = newValue
                case "wheelSlip": wheelSlip = newValue
                case "boost": boost = newValue
                case "flags": flags = newValue
                default: break
                }
            }
        }
    }

    /// Session-level numbers from a PDR file, carried into the stored blob's
    /// `meta`. `SessionMeta` is the JS's `sessionMeta`.
    public struct SessionMeta: Hashable, Sendable {
        public var ambientC: Double?
        public var intakeC: Double?
        public var elevationM: Double?
        /// The car's lifetime odometer, not this session's distance.
        public var odometerKm: Double?

        public init(
            ambientC: Double? = nil, intakeC: Double? = nil, elevationM: Double? = nil,
            odometerKm: Double? = nil
        ) {
            self.ambientC = ambientC
            self.intakeC = intakeC
            self.elevationM = elevationM
            self.odometerKm = odometerKm
        }
    }

    /// The outcome of lat-vs-distance lap recovery for a beacon-less recording.
    public struct LapRecovery: Sendable {
        public var laps: [ParsedLap]
        public var lapM: Double
        /// Periodicity confidence of the recovered lap length.
        public var r: Double
        /// Whether the boundaries were pulled onto a real start/finish by a
        /// beacon-timed recording in the same batch.
        public var anchored: Bool
        /// Template-match confidence, once anchored.
        public var phaseR: Double?

        public init(
            laps: [ParsedLap], lapM: Double, r: Double, anchored: Bool, phaseR: Double? = nil
        ) {
            self.laps = laps
            self.lapM = lapM
            self.r = r
            self.anchored = anchored
            self.phaseR = phaseR
        }
    }
}

/// A lap as a parser reports it.
///
/// `Geo.DerivedLap` with two differences that matter: a PDR beacon carries an
/// absolute crossing number (`lapNumber`), and a hand-entered lap has no time
/// window at all — which is exactly the case `buildLapChannels` skips.
public struct ParsedLap: Hashable, Sendable {
    /// The recorder's own crossing number, when there is one.
    public var lapNumber: Int?
    public var timeMs: Int
    /// Derived rather than timed by a beacon. Rendered with a `~` prefix.
    public var estimated: Bool
    /// On the telemetry clock, the same clock as `gps[].t`.
    public var startT: Double?
    public var endT: Double?

    public init(
        lapNumber: Int? = nil, timeMs: Int, estimated: Bool, startT: Double? = nil,
        endT: Double? = nil
    ) {
        self.lapNumber = lapNumber
        self.timeMs = timeMs
        self.estimated = estimated
        self.startT = startT
        self.endT = endT
    }

    public init(_ lap: Geo.DerivedLap) {
        self.init(timeMs: lap.timeMs, estimated: lap.estimated, startT: lap.startT, endT: lap.endT)
    }
}

extension ParsedRecording {
    /// A finished phone recording as a parsed source, so it enters the same review
    /// flow an imported clip does.
    public var asTelemetry: ParsedTelemetry {
        ParsedTelemetry(
            kind: .live,
            date: date,
            time: time,
            durationS: durationS,
            laps: laps.map(ParsedLap.init),
            gps: gps,
            needsLine: needsLine
        )
    }
}

File: apps/android/core/src/main/kotlin/app/trackevolution/core/telemetry/ParsedTelemetry.kt
package app.trackevolution.core.telemetry

import app.trackevolution.core.GpsPoint
import app.trackevolution.core.Lap
import app.trackevolution.core.ParsedRecording
import app.trackevolution.core.TraceSample
import app.trackevolution.core.model.SessionChannels

/**
 * One parsed telemetry source, whatever produced it.
 *
 * The shape every parser in `public/js/import/parse.js` resolves to, and the
 * reason the review flow doesn't care where laps came from: a PDR file with
 * beacons arrives with exact laps and skips the line picker, a GoPro clip
 * arrives with a trace and [needsLine], and a phone recording
 * ([ParsedRecording]) converts into the same value via [asTelemetry]. The
 * review flow in `:app` takes this so all three share one screen — the same
 * generalisation NS-30 made on iOS.
 *
 * Immutable, unlike the JS object the importer mutates in place: every step
 * ([TelemetryChannels.attachLapChannels], [Telemetry.applyGate],
 * [PDRLaps.anchorPdrBatch]) returns the value it would have produced. Same
 * names, so the two still diff by eye.
 */
public data class ParsedTelemetry(
    val kind: Kind,
    /** Local `yyyy-MM-dd` from the file, when it carries one. */
    val date: String? = null,
    /** Local `HH:mm:ss`. */
    val time: String? = null,
    val durationS: Double = 0.0,
    val laps: List<ParsedLap> = emptyList(),
    /**
     * Decoded GPS trace in degrees, or null when the source has none the app
     * can use — normal for a PDR recording whose coordinates don't decode.
     */
    val gps: List<GpsPoint>? = null,
    /** The user has to pick a start/finish line before this has laps. */
    val needsLine: Boolean = false,
    /** Best-lap polyline, drawn as the racing line on the event page. */
    val bestLapTrace: List<TraceSample>? = null,
    /** Per-lap channel arrays on a driven-distance grid, stored with the session. */
    val lapChannels: SessionChannels? = null,
    /** Session maxima from a PDR file's car channels. */
    val metrics: Metrics? = null,

    // ---- PDR-only ----------------------------------------------------------

    /**
     * How many beacon crossings the recorder actually logged. Zero means every
     * lap here was derived rather than timed.
     */
    val beaconCount: Int = 0,
    /** Raw latitude and odometer series, kept for lap recovery ([PDRLaps]). */
    val channels: RawChannels? = null,
    /** Scaled car channels for the per-lap graphs. */
    val carChannels: CarChannels = CarChannels(),
    /**
     * The slow channels, keyed by `SCALAR_NAMES`, before they are reduced to one
     * value per lap. A map rather than a data class because nothing reads an
     * individual one — `buildLapChannels` walks the whole list.
     */
    val lapScalarChannels: Map<String, List<ChannelPoint>> = emptyMap(),
    /** Session-level numbers, stored as the channel blob's `meta`. */
    val sessionMeta: SessionMeta? = null,
    /** Set when laps came from lat-vs-distance periodicity rather than beacons. */
    val lapRecovery: LapRecovery? = null,
) {
    /** Which parser produced this. `KIND_LABELS` in the JS. */
    public enum class Kind(public val rawValue: String, public val label: String) {
        PDR("pdr", "PDR"),
        GOPRO("gopro", "GoPro"),
        /** Not a file parser: the in-app lap recorder. */
        LIVE("live", "Recorded"),
    }

    /**
     * Session maxima — each null when the file doesn't carry the channel, or
     * carries one whose peak isn't plausible.
     */
    public data class Metrics(
        val topSpeedKph: Double? = null,
        val maxRpm: Double? = null,
        val maxLatG: Double? = null,
        /**
         * Peak braking, reported positive the way a driver talks about it — it
         * is the negative half of longitudinal G.
         */
        val maxBrakeG: Double? = null,
        val maxBoostKpa: Double? = null,
        val maxOilC: Double? = null,
    )

    /**
     * Session-level numbers from a PDR file, carried into the stored blob's
     * `meta`. `sessionMeta` in the JS.
     */
    public data class SessionMeta(
        val ambientC: Double? = null,
        val intakeC: Double? = null,
        val elevationM: Double? = null,
        /** The car's lifetime odometer, not this session's distance. */
        val odometerKm: Double? = null,
    )

    /** The two raw PDR series lap recovery works from. */
    public data class RawChannels(
        val latPts: List<ChannelPoint>,
        val odoPts: List<ChannelPoint>,
    )

    /**
     * Car channels in display units: speed km/h, latG in G, throttle/brake in
     * %, steering in signed steering-wheel degrees. Null where the file lacks
     * the channel.
     */
    public data class CarChannels(
        val speed: List<ChannelPoint>? = null,
        val rpm: List<ChannelPoint>? = null,
        val latG: List<ChannelPoint>? = null,
        val throttle: List<ChannelPoint>? = null,
        val brake: List<ChannelPoint>? = null,
        val steering: List<ChannelPoint>? = null,
        /** Signed: negative under braking. */
        val longG: List<ChannelPoint>? = null,
        /** Degrees per second, signed. */
        val yaw: List<ChannelPoint>? = null,
        /** 1–8, or 0 for the clutch-in / no-gear state. */
        val gear: List<ChannelPoint>? = null,
        /** (driven − non-driven) wheelspeed, percent. */
        val wheelSlip: List<ChannelPoint>? = null,
        /** Manifold gauge pressure, kPa. */
        val boost: List<ChannelPoint>? = null,
        /** ABS | traction control shl 1 | stability control shl 2. */
        val flags: List<ChannelPoint>? = null,
    ) {
        /** `chans[name]` in the JS. */
        public operator fun get(name: String): List<ChannelPoint>? = when (name) {
            "speed" -> speed
            "rpm" -> rpm
            "latG" -> latG
            "throttle" -> throttle
            "brake" -> brake
            "steering" -> steering
            "longG" -> longG
            "yaw" -> yaw
            "gear" -> gear
            "wheelSlip" -> wheelSlip
            "boost" -> boost
            "flags" -> flags
            else -> null
        }

        public fun with(name: String, values: List<ChannelPoint>?): CarChannels = when (name) {
            "speed" -> copy(speed = values)
            "rpm" -> copy(rpm = values)
            "latG" -> copy(latG = values)
            "throttle" -> copy(throttle = values)
            "brake" -> copy(brake = values)
            "steering" -> copy(steering = values)
            "longG" -> copy(longG = values)
            "yaw" -> copy(yaw = values)
            "gear" -> copy(gear = values)
            "wheelSlip" -> copy(wheelSlip = values)
            "boost" -> copy(boost = values)
            "flags" -> copy(flags = values)
            else -> this
        }
    }

    /** The outcome of lat-vs-distance lap recovery for a beacon-less recording. */
    public data class LapRecovery(
        val laps: List<ParsedLap>,
        val lapM: Double,
        /** Periodicity confidence of the recovered lap length. */
        val r: Double,
        /**
         * Whether the boundaries were pulled onto a real start/finish by a
         * beacon-timed recording in the same batch.
         */
        val anchored: Boolean,
        /** Template-match confidence, once anchored. */
        val phaseR: Double? = null,
    )
}

/**
 * A lap as a parser reports it.
 *
 * [Lap] (the geometry's derived lap) with two differences that matter: a PDR
 * beacon carries an absolute crossing number ([lapNumber]), and a hand-entered
 * lap has no time window at all — which is exactly the case
 * [TelemetryChannels.buildLapChannels] skips.
 */
public data class ParsedLap(
    /** The recorder's own crossing number, when there is one. */
    val lapNumber: Int? = null,
    val timeMs: Int,
    /** Derived rather than timed by a beacon. Rendered with a `~` prefix. */
    val estimated: Boolean,
    /** On the telemetry clock, the same clock as `gps[].t`. */
    val startT: Double? = null,
    val endT: Double? = null,
) {
    public companion object {
        /** A line-derived lap from the geometry port. */
        public fun of(lap: Lap): ParsedLap =
            ParsedLap(timeMs = lap.timeMs, estimated = lap.estimated, startT = lap.startT, endT = lap.endT)
    }
}

/**
 * A finished phone recording as a parsed source, so it enters the same review
 * flow an imported clip does.
 */
public fun ParsedRecording.asTelemetry(): ParsedTelemetry = ParsedTelemetry(
    kind = ParsedTelemetry.Kind.LIVE,
    date = date,
    time = time,
    durationS = durationS,
    laps = emptyList(),
    gps = gps,
    needsLine = needsLine,
)
